modules/partition-distribution/src/main/java/org/apache/ignite/internal/partitiondistribution/AssignmentsChain.java

```diff
@@ -64,31 +71,70 @@ public AssignmentsChain replaceLast(Assignments newLast) {
      * @param newLast New last link.
      * @return new AssignmentsChain.
      */
-    public AssignmentsChain addLast(Assignments newLast) {
+    public AssignmentsChain addLast(Assignments newLast, long configurationTerm, long configurationIndex) {
         assert !chain.isEmpty() : "Assignments chain is empty.";
 
-        List<Assignments> newChain = new ArrayList<>(chain);
+        List<AssignmentsLink> newChain = new ArrayList<>(chain);
 
-        newChain.add(newLast);
+        newChain.add(new AssignmentsLink(newLast, configurationTerm, configurationIndex));
 
         return new AssignmentsChain(newChain);
     }
 
+
+    /**
+     * Gets the next link in the chain after the given link.
+     *
+     * @param link The link to get the next one from.
+     * @return The next link in the chain, or {@code null} if the given link is the last one in the chain.
+     */
+    public @Nullable AssignmentsLink nextLink(AssignmentsLink link) {
+        int i = chain.indexOf(link);
+
+        return i < 0 || i == chain().size() - 1 ? null : chain.get(i + 1);
+    }
+
+    /**
+     * Returns the last {@link AssignmentsLink} in the chain that contains the specified node.
+     *
+     * @param nodeConsistentId The consistent identifier of the node to search for.
+     * @return The last {@link AssignmentsLink} that contains the node, or {@code null} if no such link exists.
```

Comment:
I'd also add an example from the ticket here in order to illustrate what last means.

> on input link1([A,B,C,D,E,F,G]) > link2([E,F,G]) > link3([G])
chain.lastLink(F) returns link2(E,F,G).